app/src/keymap.c

```diff
@@ -247,13 +247,18 @@ int zmk_keymap_sensor_triggered(uint8_t sensor_number, const struct device *sens
 #endif /* ZMK_KEYMAP_HAS_SENSORS */
 
 int keymap_listener(const zmk_event_t *eh) {
-    if (is_zmk_position_state_changed(eh)) {
-        const struct zmk_position_state_changed *ev = cast_zmk_position_state_changed(eh);
-        return zmk_keymap_position_state_changed(ev->position, ev->state, ev->timestamp);
+    const struct zmk_position_state_changed *pos_ev;
 #if ZMK_KEYMAP_HAS_SENSORS
-    } else if (is_zmk_sensor_event(eh)) {
-        const struct zmk_sensor_event *ev = cast_zmk_sensor_event(eh);
-        return zmk_keymap_sensor_triggered(ev->sensor_number, ev->sensor, ev->timestamp);
+    const struct zmk_sensor_event *sensor_ev;
+#endif /* ZMK_KEYMAP_HAS_SENSORS */
+
+    if ((pos_ev = as_zmk_position_state_changed(eh))) {
+        return zmk_keymap_position_state_changed(pos_ev->position, pos_ev->state,
+                                                 pos_ev->timestamp);
+#if ZMK_KEYMAP_HAS_SENSORS
+    } else if ((sensor_ev = as_zmk_sensor_event(eh))) {
+        return zmk_keymap_sensor_triggered(sensor_ev->sensor_number, sensor_ev->sensor,
+                                           sensor_ev->timestamp);
```

Comment:
Can simplify this to

```c
const struct zmk_position_state_changed *pos_ev = as_zmk_position_state_changed(eh);
if (pos_ev) {
    return ...;
}
// no need for "else" because of return above.
#if ZMK_KEYMAP_HAS_SENSORS
const struct zmk_sensor_event *sensor_ev = as_zmk_sensor_event(eh);
if (sensor_ev) {
    return ...;
}
#endif
```

Also, it looks like this function can reach the end without returning anything? I think that's undefined behavior, but I'm not really sure on that. I usually have the warning for missing return treated as an error.

Reply:
Indeed, that is much nicer. There is a final return at the end of the function already though?

Reply:
Sorry. I think I was getting confused because GitHub stopped the context right after the closing brace on line 258 of the original code, so it looked like the end of the function.